### ISSUE TITLE
Question: adopt some of the svelte template syntax?

### ISSUE THREAD

**Issue by evs-chris** [ractivejs/ractive#2790]
## Description:
As a sort of sister (great-granddaughter?) of Ractive, [svelte](https://github.com/sveltejs/svelte) has started off life looking quite similar to Ractive. It does have some slightly different syntax regarding bindings and directives, which I quite like. Here's the summary:

| Ractive | Svelte |
| --- | --- |
| `on-click="..."` | `on:click="..."` |
| `value="{{...}}"` | `bind:value="..."` |
| (component mappings) `someVar="{{...}}"` | `bind:someVar` === `bind:someVar="someVar"` |
| `fade-in`, `slide-out`, `fade-in-out="..."` | not settled, but probably `in:fade="..."`, etc |
| `as-foo`, `as-bar="..."` | also not settled, but probably `as:foo="..."` |
| `class-whatever="{{...}}"` | probably not to be included, but I'd use `class:whatever="..."` |
| `style-display="{{...}}"` | same as class - `style:display="..."` |

Svelte also has a dedicated `ref:name` that creates a reference on the component named `name` that points to the element. Ractive does automatic id collecting, though I'm about to propose that be removed for 0.9, but I think that's best left to a decorator `<div as:ref="'joe'" />` and let the ref decorator decide where to stow `joe`. I'm not crazy about the nested quotes there, but it's basically required for parsing purposes, as otherwise the parser will assume it's a ref and try to look it up on the viewmodel.

I've never really been fond of the `bind:` syntax, but it would resolve some issues we've had with users not being able to tell the difference between an attribute (byval) and a mapping (byref) on components.  With explicit binding on mappings, components could more easily _require_ that data be passed in a certain way, since as it stands, a non-bindable (computed) mapping ends up just being passed by value even if it looks like a binding.

Adding that prefix would also allow the removal of the mustaches required for two-way binding, which would all us to unify _all_ of the directives syntactically. Then any attribute with mustaches would just be string interpolation with `setAttribute`, and anything directivized would act according to the particular directive (but not require mustaches). It would also eliminate a fair amount of special casing in the code, though a few, like (only? because it changes the element to a completely different animal - a checkbox, a select, and a text field have little in common) `type` on inputs, would remain.

Having `bind:value="foo" value="bar"` would also allow the binding to initialize with whatever is in the actual value attribute.

## Versions affected:
0.9.0/edge or 0.10.0

**Comment by ghost** [comment]
Yep, was just thinking about this when it was announced. However...

I was leaning towards how Vue does it tho, the `v-command:name="something"`. Feels much more natural and consistent. It has `v-` prefix that makes it stand out. We can have ours be like `r-`.

```xml
 <input type="text" r-bind:value="expression|data|etc.">

 <span r-decorate:tooltip="{ foo: bar, baz: bam }">
 <span r-as:tooltip="{ foo: bar, baz: bam }">

 <span r-transition:fly="{ foo: bar, baz: bam }">

 <span r-style:display="inline-block">
 <span r-style:display="isMobile ? 'block' : 'inline-block'">

 <span r-class:active="true">
 <span r-class:active="isActive && !isDisabled">

 <button r-on:click="doSomething()">

 <Component r-bind:name="foo" r-bind:address="bar" r-on:customevent="doSomething()"/>
```

Wordy, true, but definitely less confusing and more consistent. 

> Adding that prefix would also allow the removal of the mustaches required for two-way binding, which would all us to unify all of the directives syntactically. Then any attribute with mustaches would just be string interpolation with setAttribute, and anything directivized would act according to the particular directive (but not require mustaches)

👍 on this one. This will make things feel a bit like Angular.

- Anything inside an `r-` attribute is subject to evaluation, like Angular's `ng*`. Can be a reference, expression, DSL (like in its repeat that looks like Python comprehensions) etc.

- `{{ }}` becomes a string that doesn't bind back. Appended appropriately
    - As string, i.e. `<img src="{{ imagePath }}">`
    - As textNode, i.e. `<span>{{ label }}</span>`
    - `<input value="{{ text }}"` updates value, but doesn't send changes back to `text`.

- `{{{ }}}` become DOM elements, like normal.

This also gives way to mustaches to become more of a template control mechanism (conditions, loops, wrappers, positioning, partials, yields, etc) and less to do with actual data binding except for the cases mentioned above.

> Ractive does automatic id collecting, though I'm about to propose that be removed for 0.9

👍 for removal with no replacements. We have the `ractive.find*` API anyways.

my 2c.

**Comment by kouts** [comment]
For me, template syntax was one of the things I liked when choosing Ractive as my go-to MVVM library for my projects, and it's still one of the things I love about it today.

- I like to feel I'm writing (and viewing) HTML so the syntax attr="  " comes more naturally over  e.g attr:something="  "
- I prefer as-toolptip over r-decorate:tooltip any day, less code, less clutter!

**Comment by ghost** [comment]
@kouts Yep. The first bullet point definitely is the main selling point of Ractive. And I still prefer it over other frameworks because nothing comes close to this simplicity. However, consider the following:

```
data-target
on-click
contenteditable
as-tooltip
onload
fade-in-out
checked
```

Nothing really makes Ractive attributes stand out from regular attributes. Given an opportunity to do a syntax overhaul, making Ractive's special attributes stand out is a plus to readability. Of course, the trade-off is longer code. 😄

**Comment by dagnelies** [comment]
Hi,

sorry to be the party pooper again, but I'd vote against it for the simple reason that the disadvantages outweight the benefits.

The disadvantages: it breaks all existing templates (again?!)
The benefits: some find it nicer, others don't. Personally, I'd say adding `bind:` everywhere just make it look more verbose, that's about it.

----

What I'd really love to see instead is an *easy* way to make widgets with inheritable attributes and some stability. But currently, half of the plugins are broken from version to version and poorly maintained, including ractive's own plugins.

Btw, I've never seen the `as-...`, `class-...` and `style-...`. Where are they documented? Why not use `class="{{...}}"` and `style="{{...}}"` as before?

**Comment by kouts** [comment]
+1 @dagnelies is right, probably the biggest issue is that this breaks all existing templates.
As much as I dislike the "breaking" part, I must say that reading through @evs-chris post I understand
that there will be a reduction in Ractive's size and maybe a small performance gain?

For this to happen is it necessary that the change must follow the double colon paradigm?
Could it be something as the decorators is now?

e.g something along the lines of:
on-click="..." -> r-onclick=""
value="{{...}}" -> r-value=""

**Comment by aharpervc** [comment]
👎 on this overall, I actually hate vue's prefixed binding attributes and prefer ractive's more natural event and binding syntax. In fact I was just this week discussing ractive vs vue with a coworker now that vue 2.0 is out (reviewing the landscape) and ractive again came out on top, and vue's awful directive syntax is part of that.

And 👎 x 1000 for getting rid of mustache binding. Why is this even coming up? As @dagnelies pointed out, this will break every template. A change that breaks every template must have an incredible, unbeatable upside. And that upside is: similarity to other libraries I guess? That is not a good calculus.

This all feels like change for change sake. If I wanted to write vue I'd be writing vue. If I wanted to be writing svelte I'd be writing svelte. Just let ractive be ractive.

**Comment by evs-chris** [comment]
__TL:DR__ The aesthetics (`-` vs `:` and prefixing) of the change don't _really_ matter to me, it's more the ambiguity that can be removed and the simplification of internals that result, features that could be exposed more coherently, and corner cases that could be addressed where they currently cannot. Though I certainly do have an opinion on aesthetics too :smile:

@dagnelies @aharpervc But my goal is to have a major breaking change for each release :grinning: Kidding. I don't enjoy breaking changes any more than anyone else, but this is the main reason I want to keep more than one actively maintained version. So no one is forced into updating an entire app because bugs have surfaced in their now-unsupported version. I think Ractive still has a bit of evolution to do, but I don't want anyone to get burned along the way.

I specifically do _not_ like vue directives. `v-for` and `v-if` in particular annoy me. Vue is clearly a lovely framework that does many nice things, but it just doesn't fit with me aesthetically or functionally. Conditionals and iteration _shouldn't_ be element directives. Aside from that, I also don't like the letter-hyphen prefix. It feels clutter-y to me, where having the namespace (binding directive, event, decorator, transition) is sufficient to convey what is going to happen.

I posed this question, because internally Ractive's method of handling bindings and related directives suffers from the various potential ambiguities that the syntax allows. We've had issues where, for instance, name binding completely trashes the name attribute on form elements, so the forms can't be submitted using a good old-fashioned form post. Having `bind:name="list" name="myFormList"` or even `name="{{ some.dynamicName.whyNot }}"` works around that and other similar issues. The `-` to `:` is more of an aesthetic change, because I thought `bind:value` looks better than `bind-value`.

Moving to a namespace-style binding would also open up binding directives inline with what they _actually_ apply to. For instance, `bind:lazy(500):value="query"` could be used for a lazy value binding that debounces at 500ms. `in:name(currentTransitionName)="{ args: ... }"` could be used to get dynamic transitions and decorators back in a cleaner and more unified way. I've also been looking for a way to add more control, like #2545 and the related PR request, and binding flags like that would allow it without a bunch of extra special parsing and handling. It also grants a nice shortcut for components with `bind:foo` being the same as `foo="{{foo}}"`. It would simplify plain old attribute handling for a potential speed boost, because anything that needs to be looked at for binding purposes is already clearly marked as such. There's no "does this have a fragment, that only has one child, and is that child an interpolator, and does that interpolator only refer to a single keypath not a computation?" and other similar logic to run through. It also allows the mustache to be dropped out of the value side, as it's _always_ a data context and _never_ just a string. Currently, an erroneous space in your binding can cause it to be ignored e.g. `value=" {{foo}}"` because that's a legitimate bit of template that will only ever result in a string attribute value that can't be bound.

@dagnelies, yes they are documented, though the documentation is not exactly in the best shape right now. That's pretty high up on the list of things to fix before 0.10, but I'm really lousy at docs and coherence, as you may or may not be able to tell from this message :smile:. `as-whatever` is the new decorator syntax that has replaced `decorator="whatever"`, `class-foo="{{condition}}"` is a faster shortcut for adding and removing the `foo` class based on the truthiness of `condition` (an alternative for `{{#if condition}}class="foo"{{/if}}` that is composable without leaving empty or improperly spaced class attributes on the element), and `style-display="{{something}}"` is a faster shortcut for setting a particular style property. `class` and `style` got special handling in 0.8 so that they don't break or get broken by transitions, decorators, or other outside interference.

**Comment by tomekmarchi** [comment]
Quick blunt 2 cents.

- I hate Vue syntax. The JS API (stupid shit like $ signs (Terrible idea mongo is guilty of this too) ) and the HTML has just down right none sensical lexical syntactical sugar layer of pointless pseudo-complexity.
- changing a character that is standard HTML - to : is just syntactical sugar as you said. The closer it is to HTML the more it feels cohesive and right.
- Maybe there is extra functionality that we could take advantage of by adding ":" into the syntax now that's a thought. Maybe a sort of lazy way to do something extra. Instead of just swapping the character.
- I find svelte interesting but so far what they propose is not enough to warrant a thought of change.  In my eyes the biggest issue with Ractive, that is being worked out, is simply put it's size. It needs to move into a post IE world IMO. Which it is doing. After that is taken care of it's pretty close to perfect in my eyes. Yeah we can work on some dynamic cases and under the hood speed improvements. However, overall Ractive is exactly what I need and how I see a view library should be.
- In conclusion Ractive is as close to diamond as you can get. It's an iconic library. React & Vue seem more of just another kid playing with things that aren't broken for the sake of being "different" or "edgy". Let's not fall into that trap of dipshittery.

**Comment by aharpervc** [comment]
A single thumbs up emoji is not enough to express how much I agree with @tomekmarchi

**Comment by evs-chris** [comment]
To be fair, `:` is also standard, as it's used for namespaced elements and attributes e.g. `<use xlink:href="..." />` in embedded svgs. Again, not that it really matters to the parser :smile:, though it could possibly eliminate some identifier issues as `:` is _never_ valid in an identifier where `-` sometimes is e.g. `on-click-cusom\-event` vs `on:click:custom-event`. Parsing escapes without lookbehind support in your regexes is not pretty.

Ignoring the separator for now, which can certainly remain `-`, does that mean that everyone is pretty well against `bind-value="foo"` replacing `value="{{foo}}"`? Again, the nice thing about the prefix is that it lets the reader _and_ ractive know that it's a directive and not a normal attribute - allowing errors to move to parse time rather than run time for those who pre-parse. It would also eliminate the need for a twoway flag - one way: `value="{{foo}}"`, two way: `bind-value="foo"`.

What about moving binding flags from attributes inline to the binding e.g. `bind-value-lazy="foo"` or `value-lazy="{{foo}}"` with optional params `value-lazy(500)="{{foo}}"` as opposed to `value="{{foo}}" lazy` and `value="{{foo}}" lazy="500"`? Establishing flags as part of the binding that they apply to opens up some possibilities that are not pretty to implement right now. Also, my example above should have followed a more uniform `directive:primaryArg:flags`, so `bind:value:lazy(500)="someRef"` or in the kebab flavor `directive-primaryArg-flags` for `bind-value-lazy(500)="someRef".

I suppose I need to scrap this issue, write up my questions coherently along with the benefits associated with each potential change, and open a new issue. Looking back my previous posts (starting with this one), it looks like I may have been intoxicated :laughing:

**Comment by dagnelies** [comment]
Well, from my user perspective, what I like most about ractive, is it's more or less simplicity.
Placing `{{foo}}` anywhere where you want foo to be is just that. It's simple, it works, it's good. IMHO, anything else will just increase the cognitive overhead and would be bad.

> Again, the nice thing about the prefix is that it lets the reader and ractive know that it's a directive and not a normal attribute 

...see, that's my point. Why would a user want to learn the intricacies of all that? It's close to internal mechanics. You shouldn't. If I have a `{{foo}}` in my template, it should appear there. End of line. And if it's a value, then it should change according to user's input. Simple, straightforward, and everything you need.

**Comment by tomekmarchi** [comment]
@evs-chris I do see that perspective when I ponder that question a bit more. I would be interested in extending the syntax possibly with ":". Do you have any ideas as to what we could do with it to make developers lives easier?

**Comment by tomekmarchi** [comment]
@evs-chris Want to put this svelte thought to bed. I would say the majority of cases using Ractive, Vue and React would not benefit from nor enjoy Svelte at it's current iteration. Here is why. It has taken the perspective that developers in the future will be using a build process for sites. The way I see it is this. We in fact will be using build processes a lot less because the barebones tooling of the environment and the increase in productivity itself will result in less bundling and files. Not only that but when we look at the way the web is progressing apps are becoming larger, more complex, highly isomorphic, utterly dynamic. Therefore I see it as being counter intuitive to long term development projections. I anticipate importing things on the fly only when needed. Meaning it's best to actually have many organized small files (modules) separately than bundling or compiling initially. The lazy web has two main advantages that's easier development but also drastically improved application performance. The philosophy of what you need only exactly when you need it; I see as the web's only viable future. Ractive while it decreases with size over time and adopts new technologies will be here for a while. JSX will die and so will crappy syntax like having $ in your API.

Svelte complains about the same sickness of too large of files being served to the client. 
"We're shipping too much code to our users."

I **agree**. **HOWEVER** their prescription for this virus is entirely off **mark**. _Foolishly_ so. "Svelte compiles them into tiny standalone JavaScript modules." This one sentence forever will condemn the fate of Svelte. They had a point in the beginning of the blog post then shit the bed when it came to a solution for the problem. It totally missed the entire direction the web is heading when we take into account import and components. These two little functions hint at the future of the web. It's a huge hint a huge clue that says "when you need it only when you need it". When I read that one word "compiles" I knew in the deepest part of my soul that they "done fucked up".

**Comment by evs-chris** [comment]
I can see it from both sides. `<input value="{{whatever}}" />` is beautiful and dead simple syntactically. The only mildly challenging thing is the mustache, but once you grok those, you've pretty much got it. More advanced usages are where things start getting, shall we say, less beautiful. I also think there's a fair amount of usability win in having the same context for all directives e.g. no mustaches required because you're already in data mode inside the quote (or after the `=` if you're one of those heathens that doesn't quote attributes :wink:). So I guess (for me):

Pros:
  * more unified, and somewhat simplified syntax
  * less fighting with escapes
  * better support for binding corner-cases
  * drop a flag that can be confusing (twoway)
  * binding flags can be inlined and expanded more readily to e.g. handle passing custom comparators for syncing up multi-selects and whatnot
  * a bit of speedup in element rendering because if there's no `bind` directive, there's no need to even start the check for a twoway binding
  * dynamic directive names can come back without getting weird

Cons:
  * breaking template structure __and__ template syntax, both of which are already done on edge due to deprecation removals
  * lose the beautiful simplicity for the common case
  * I'm not wild about the transition case, because `fade-in` and `slide-in-out` are pretty awesome visually, where `in:fade` and `in-out:slide` are less so. I want to be able to add chaining to transitions when they get their refactor, and this is one of the things I've considered: `<div in:grow:order(1) in:highlight:order(2)="{ duration: 1000, color: 'lightoffpaleblue' }">`, which would `grow` the element in and then flash the background color to `lightoffpaleblue` and back to start. Actually, looking at it now, it might make sense to move all of the standard transition params to flags `in:highlight:order(2):duration(1000):color('lightoffpaleblue')`. At this point, in my head, the implementation of directive flags results in `{ [flagName]: [ ...flagValues ], ... }`, so it would simplify the implementation of transition plugins a bit too.

**Comment by evs-chris** [comment]
FWIW, Ractive won't be moving toward the compilation approach take by svelte. I think the compilation approach makes a ton of sense for certain things and the runtime lib approach makes a ton of sense for other certain things. I only mentioned svelte because the fresh perspective on template syntax got me thinking about some of the cruftier code in Ractive and how it may be able to be improved.

**Comment by evs-chris** [comment]
Just thought of another potential pro for directive flags: event delegation. It wouldn't be hard to have an event directive flag that told it to use the context of the target element, rather than the element with the directive, when processing.

**Comment by dagnelies** [comment]
...well, I hope you don't miss the fact that people use ractive *because* it is simple, and *because* it uses the familiar mustache syntax. Remove this and there will be little reasons to pick up ractive at all, rather than some more popular alternatives like angularjs.

**Comment by kouts** [comment]
What @dagnelies said pretty much sums it up for most Ractive users I think, it sure does for me.
Easy switching from "plain ol static mustache templates" was why I picked Ractive, sometimes it's the seemingly small things that matter.

**Comment by dagnelies** [comment]
Thank you @evs-chris !

**Comment by sarendipitee** [comment]
Been following this thread, but haven't chimed in yet... just wanted to add that as someone who maintains a dozen or so open source ractive components, changing template syntax again will make me cry :(

**Comment by arxpoetica** [comment]
Agree w/ @dagnelies. Don't break something that is simple and works. As much as I love Svelte, I don't like the new syntax at all. :(

**Edit:** I take back my comment about not liking the syntax. It helps a beginner know this isn't standard HTML, i.e., it's a framework/compiler doing something.

**Comment by evs-chris** [comment]
Don't worry, it's not going anywhere 😄. I did add `bind-` prefixed attributes that have values in an expression context rather than a string context, but that's about all that came out of this issue.

**Comment by paulie4** [comment]
I just noticed that there's a `bind-` directive when I was trying to debug why my `bind-to-id` attribute was getting converted to `to-id` and found this issue about it. Can we add that to the https://ractive.js.org/api/#directives documentation, since I'm not exactly clear one what that directive is for (is it just for Ractive components?), and it will help people avoid the gotcha of trying to use it in a regular attribute name?

**Comment by PaulMaly** [comment]
If we can't use standard things like "onclick" instead of "on-click" (actually, why not? ), but if it impossible, so I thing xml-like syntax "on:click" looks nice. It looks right, because, Ractive's attributes extend standard things, like in xml.

p/s Vue's syntax is very bad.
